### ISSUE TITLE
MicroMasters Email

### ISSUE THREAD

**Issue by pdpinch** [mitodl/micromasters#694]
_Engineering doc (discusses milestones, technical specifics, etc.): https://docs.google.com/document/d/1v-DfmVxdamAakjIZFVJ-5ZsHlCaCzZeCu2ovGLUBU5Y/edit?usp=sharing_
## Milestone 1

As a MM team member, when I finish a search, I want to send an e-mail to all the students in result set. After I've sent an e-mail, I need to know that it's been delivered and how many students have opened it. I'd also like to be able to resend the message to users that match a search query, but haven't yet received it.
### MVP
- [x] ~~#726 Investigate using Mailgun to deliver emails from staff to learners~~
- [x] ~~#733 allow user to change opt-in settings~~ 
- [x] #807 configure Mailgun and #727 set up sub-domain
- [x] #831 Create a Mailgun account for MIT ODL
- [x] #981 Use batch sending for learners search (fix)
- [ ] #830 Create models to support sent-mail view
- [x] #718 API for piping mail to mail service (Mailgun)
- [ ] #832 API for fetching mail stats from Mailgun
- [x] #719 Component for composing (text) e-mail
- [ ] #808 sent-mail view
- [ ] #809 allow staff to resend mail
- [x] #805 basic mail template

Undecided:
- [ ] #958 notifying a user about the success/failure of their sent email
## Possible issues for future versions
- [x] allow staff to send HTML mail (requires #717 discovery)
- [ ] e-mail inbox view
- [ ] support inbound mail
- [ ] notify users about course enrollment deadlines

**Comment by gsidebo** [comment]
I created an eng doc to outline some technical and feature-related decisions to be made in regard to email/messaging in MicroMasters in general: https://docs.google.com/a/mit.edu/document/d/1hc8RH3Rn85hvWVUr6nfrO88NnskPs0C5o6KxN7_m8D8/edit?usp=sharing _(note that this link is different from the one I originally sent out via email)_

That eng doc addresses our needs for email/messaging in the future, but this particular story can and will be much more limited in scope. After a conversation with @Ferdi , we figured out that **this story should center around enabling an instructor to send one-way, text-only, non-reply-able emails to learners.** 

That means that, for this issue, we won't be worrying about mail routing for replies delivered back to our app, as the description indicates. I've done some [evaluation of Mailgun](https://docs.google.com/document/d/1hc8RH3Rn85hvWVUr6nfrO88NnskPs0C5o6KxN7_m8D8/edit#heading=h.llwvcpc5figm), and that functionality is indeed available to us, but it won't be a part of this issue.

**Comment by gsidebo** [comment]
If the goal of this issue was to evaluate Mailgun for our current messaging needs (one-way instructor-to-learner communication) and future needs, I'm giving it the thumbs up pending a discussion with devops (just to make sure that there are no technical barriers). I'll leave it to @pdpinch and @Ferdi to decide if the [pricing](http://www.mailgun.com/pricing) is reasonable. As I mentioned in the [eng doc](https://docs.google.com/document/d/1hc8RH3Rn85hvWVUr6nfrO88NnskPs0C5o6KxN7_m8D8/edit#heading=h.llwvcpc5figm), it's free up to 10,000 messages a month, and each email is $0.00050 after that. 50,000 emails in a month comes out to $20.

I think that looking into a "solution for composing messages (possibly in app; look at Ink from Foundation)" is a separate concern and should be explored as part of a separate issue. Mailgun satisfies all of the criteria listed in the description.

**Comment by gsidebo** [comment]
@blarghmatey says everything looks pretty straightforward for Mailgun integration. I'm going to create these new issues: 

1. Looking into Foundation for email encoding and basic templating (we'll probably want to get a better idea of what we want in terms of templates)
1. Actually implementing the email functionality for users in a results set (might become 2 stories; 1 for the composition on the front end, 1 for creating an endpoint in MM to send the message)

I will also go ahead and set up an ODL account with Mailgun (@blarghmatey gave me the go-ahead). Even if we don't end up going with it, it only costs money when you go over 10k messages in a month

**Comment by Ferdi** [comment]
@gsidebo Anything we need to do regarding their suggestions ? https://documentation.mailgun.com/best_practices.html

**Comment by gsidebo** [comment]
@Ferdi most of that document concerns marketing email practices, and many of the possible issues are taken care of due to the fact that we're part of MIT. the one thing we might consider is creating a new subdomain for transactional emails coming from MicroMasters, and it's mentioned in other parts of their documentation. i'll discuss the need for it with ops, but if we decide to do that, it will be pretty straightforward

**Comment by pdpinch** [comment]
@gsidebo I want to discuss the scope of this with you, but in the meantime, I'm creating a new issue to capture the discovery work that you did on Mailgun. I'm changing this issue into an epic and will add the other bits and pieces.

**Comment by pdpinch** [comment]
I updated the description to better match our current thinking about this epic. @gsidebo and @roberthouse54 let me know if I'm off-base anywhere.

**Comment by gsidebo** [comment]
@pdpinch i made some sorta-organizational changes. looks good in general. one change i want to make: i think the data modeling should be captured in a single issue (not one issue for sent-mail models and another for groups). as far as groups are concerned, i think we should limit the scope of that model to the features we want for milestone 1 until we have a concrete use case for something more. if we want a 'fixed'/'manifestation' group concept down the road, we can add that without a ton of effort, but i think we should limit the scope for now

**Comment by roberthouse54** [comment]
I personally think that the concept of "groups" is something that could use some high level brainstorming, as this is an area where we can possibly come up with innovative feature ideas. The basic idea is a group of users. But how this works is unclear. It can be like a facebook group, a walled off community of people with a suite of features, or it can be more like a mailing list, or a group within the context of forums, or chat (like a chat room). Groups can also be created automatically, for example, the group of users that is in a particular course run. I think we have largely focused on the idea of creating "custom groups", but groups conceptually, is a bit different, in my opinion. In a sense, it is just some set of users that can collaborate, share information, etc.

**Comment by gsidebo** [comment]
a group with a suite of features is veering into roles/permissions territory, which is something we're already building separately, and which shouldn't be mixed into the other ideas that you're offering (IMO). right now, our only defined use case for this concept is just groups of people that match an instructor's search query, but i agree that we should think about broadening that concept. an instructor may want to define groups arbitrarily on their own

**Comment by pdpinch** [comment]
After speaking with @roberthouse54 I agree that groups are out of scope for this epic.

**Comment by gsidebo** [comment]
i arranged the Milestone 1 issues into a roughly sensible order based on interdependency. here are some notes ahead of the sprint starting 8/15:

- #830 (models) can be started right away
- #718 (API to send mail) depends on #830 to save data about the email and recipients, but the actual communication with Mailgun just depends on us having an account (#831) or the developer setting up a test account
- #832 (API for mail stats) - like the above, this depends on us having an account (#831) or the developer setting up a test account
- #719 (Email composition component) depends on #718, but the front-end can be worked on independently
- #808 (sent-mail view) - lots of dependencies, but front end can be worked on independently

**Comment by gsidebo** [comment]
just added #981 to change the invocation of the mailgun api for learner search result emails

**Comment by pdpinch** [comment]
Since we've been operating in the current state for several months, it looks like we overshot in defining "MVP" for this feature. 

I'm going to close this epic, but leave the remaining items open.
